src/test/java/com/uid2/admin/managers/KeysetManagerTest.java

```diff
@@ -0,0 +1,227 @@
+
+    @BeforeEach
+    public void setupMocks() {
+        mocks = MockitoAnnotations.openMocks(this);
+        when(keysetProvider.getSnapshot()).thenReturn(keysetSnapshot);
+    }
+
+    @AfterEach
+    public void teardown() throws Exception {
+        mocks.close();
+    }
+
+    protected void setKeysets(Map<Integer, Keyset> keysets) {
+        keysetSnapshot = new KeysetSnapshot(keysets);
+        when(keysetProvider.getSnapshot()).thenReturn(keysetSnapshot);
+    }
+
+    @Test
+    public void testAddOrReplaceKeyset() throws Exception{
+        KeysetManager keysetManager = new KeysetManager(keysetProvider, keysetStoreWriter,
+                keysetKeyManager, true);
+
+        Map<Integer, Keyset> keysets = new HashMap<Integer, Keyset>() {{
+                put(1, KeysetManager.createDefaultKeyset(3, 1));
+                put(2, KeysetManager.createDefaultKeyset(4, 2));
+                put(3, KeysetManager.createDefaultKeyset(5, 3));
+                put(4, KeysetManager.createDefaultKeyset(6, 4));
+            }};
+
+        setKeysets(keysets);
+        final int keysetId = 5;
+        // add new keyset
+        Keyset keyset1 = KeysetManager.createDefaultKeyset(7, keysetId);
+        keysetManager.addOrReplaceKeyset(keyset1);
+        assertTrue(keysets.containsKey(5));
+        assertTrue(keyset1.equals(keysets.get(5)));
+        verify(keysetStoreWriter).upload(mapOfSize(5), isNull());
+
+        //Reset between tests so verify works
+        reset(keysetStoreWriter);
+
+        // add existing Keyset
+        Keyset keyset2 = new Keyset(keysetId, 7, "newKeyset", Set.of(1, 2, 3), 1 , true, true);
+        keysetManager.addOrReplaceKeyset(keyset2);
+        assertTrue(keysets.containsKey(5));
+        assertFalse(keyset1.equals(keysets.get(5)));
+        assertTrue(keyset2.equals(keysets.get(5)));
+        verify(keysetStoreWriter).upload(mapOfSize(5), isNull());
+    }
+
+    @Test
+    public void testCreateKeysetForClient() throws Exception {
+        KeysetManager keysetManager = new KeysetManager(keysetProvider, keysetStoreWriter,
+                keysetKeyManager, true);
+
+        Map<Integer, Keyset> keysets = new HashMap<Integer, Keyset>() {{
+            put(1, KeysetManager.createDefaultKeyset(3, 1));
+            put(2, KeysetManager.createDefaultKeyset(4, 2));
+            put(3, KeysetManager.createDefaultKeyset(5, 3));
+            put(4, KeysetManager.createDefaultKeyset(6, 4));
+        }};
+
+        setKeysets(keysets);
+        int nextId = keysetManager.getNextKeysetId();
+        // Sharer makes an empty list
+        ClientKey sharer = new ClientKey("", "", "",  "", Instant.now(), Set.of(Role.SHARER), 7, false);
+        keysetManager.createKeysetForClient(sharer);
+        Keyset sharerKeyset = keysets.get(nextId);
+        assertEquals(sharerKeyset.getAllowedSites(), Set.of());
+        // Generator makes a null list
+        nextId = keysetManager.getNextKeysetId();
```

Comment:
Can we make createKeysetForClient actually return the keyset id that was created?
We shuouldn't need to call getNextKeysetId() here, and ideally that method should be private